{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2d59045c-7b41-4f0f-9fbe-3d01b43d573b",
   "metadata": {},
   "source": [
    "## Schema Guided Prompt Generation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "61101e0b-76f5-4ebb-a56b-d964b2ffa956",
   "metadata": {},
   "outputs": [],
   "source": [
    "import yaml\n",
    "\n",
    "def generate_sparql_prompt_from_yaml(yaml_str, user_question, named_graph_uri):\n",
    "    schema = yaml.safe_load(yaml_str)\n",
    "\n",
    " \n",
    "    prefixes = schema.get(\"prefixes\", {})\n",
    "    prefix_lines = []\n",
    "    for prefix, uri in prefixes.items():\n",
    "        if isinstance(uri, dict):\n",
    "            uri_value = uri.get(\"prefix_prefix\", uri.get(\"prefix_reference\", \"\"))\n",
    "        else:\n",
    "            uri_value = uri\n",
    "        prefix_lines.append(f\"PREFIX {prefix}: <{uri_value}>\")\n",
    "    prefixes_block = \"\\n\".join(prefix_lines)\n",
    "\n",
    "    print(\"*\"*100)\n",
    "    print(prefixes)\n",
    "    print(\"*\"*100)\n",
    "    \n",
    "    summary_lines = []\n",
    "    if \"classes\" in schema:\n",
    "        for cname, cdata in schema[\"classes\"].items():\n",
    "            summary_lines.append(f\"- Class: {cname}\")\n",
    "            for slot in cdata.get(\"attributes\", {}):\n",
    "                summary_lines.append(f\"  - Slot: {slot}\")\n",
    "    schema_summary = \"\\n\".join(summary_lines)\n",
    "\n",
    "    print(\"*\"*100)\n",
    "    print(prefixes)\n",
    "    print(\"*\"*100)\n",
    "\n",
    "    \n",
    "    prompt = f\"\"\"You are a SPARQL assistant that converts user questions into SPARQL queries using RDF data based on the provided LinkML schema.\n",
    "\n",
    "            ### Graph Context:\n",
    "            All queries should use the named graph:\n",
    "            GRAPH <{named_graph_uri}> {{ ... }}\n",
    "            \n",
    "            ### Prefixes (from schema):\n",
    "            {prefixes_block}\n",
    "            \n",
    "            ### RDF Type Encoding:\n",
    "            Entity types use:\n",
    "            ?entity biolink:category \"bican:<ClassName>\"^^xsd:anyURI\n",
    "            \n",
    "            ### LinkML Schema:\n",
    "            {schema_summary}\n",
    "            \n",
    "            ### Instructions:\n",
    "            1. Use classes and attributes from the schema to interpret the user question.\n",
    "            2. Write a SPARQL query that retrieves appropriate answers.\n",
    "            3. Return only the query, wrapped in triple backticks.\n",
    "            \n",
    "            ### User Question:\n",
    "            \"{user_question}\"\n",
    "            \n",
    "            ### SPARQL Query:\n",
    "            \"\"\"\n",
    "    return prompt\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d64677e7-715f-4eb5-9dd1-9cc6eea1d110",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "****************************************************************************************************\n",
      "{'linkml': 'https://w3id.org/linkml/', 'bican': 'https://identifiers.org/brain-bican/vocab/', 'schema': 'http://schema.org/', 'ncbi': 'https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Info&id=', 'NCBIAssembly': 'https://www.ncbi.nlm.nih.gov/assembly/'}\n",
      "****************************************************************************************************\n",
      "****************************************************************************************************\n",
      "{'linkml': 'https://w3id.org/linkml/', 'bican': 'https://identifiers.org/brain-bican/vocab/', 'schema': 'http://schema.org/', 'ncbi': 'https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Info&id=', 'NCBIAssembly': 'https://www.ncbi.nlm.nih.gov/assembly/'}\n",
      "****************************************************************************************************\n",
      "You are a SPARQL assistant that converts user questions into SPARQL queries using RDF data based on the provided LinkML schema.\n",
      "\n",
      "            ### Graph Context:\n",
      "            All queries should use the named graph:\n",
      "            GRAPH <https://www.brainkb.org/version01> { ... }\n",
      "            \n",
      "            ### Prefixes (from schema):\n",
      "            PREFIX linkml: <https://w3id.org/linkml/>\n",
      "PREFIX bican: <https://identifiers.org/brain-bican/vocab/>\n",
      "PREFIX schema: <http://schema.org/>\n",
      "PREFIX ncbi: <https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?mode=Info&id=>\n",
      "PREFIX NCBIAssembly: <https://www.ncbi.nlm.nih.gov/assembly/>\n",
      "            \n",
      "            ### RDF Type Encoding:\n",
      "            Entity types use:\n",
      "            ?entity biolink:category \"bican:<ClassName>\"^^xsd:anyURI\n",
      "            \n",
      "            ### LinkML Schema:\n",
      "            - Class: GeneAnnotation\n",
      "  - Slot: referenced_in\n",
      "- Class: GenomeAnnotation\n",
      "  - Slot: reference_assembly\n",
      "- Class: GenomeAssembly\n",
      "- Class: AnnotationCollection\n",
      "  - Slot: annotations\n",
      "  - Slot: genome_annotations\n",
      "  - Slot: genome_assemblies\n",
      "            \n",
      "            ### Instructions:\n",
      "            1. Use classes and attributes from the schema to interpret the user question.\n",
      "            2. Write a SPARQL query that retrieves appropriate answers.\n",
      "            3. Return only the query, wrapped in triple backticks.\n",
      "            \n",
      "            ### User Question:\n",
      "            \"List all genes with their molecular type\"\n",
      "            \n",
      "            ### SPARQL Query:\n",
      "            \n"
     ]
    }
   ],
   "source": [
    "with open(\"genome_annotation.yaml\") as f:\n",
    "    yaml_data = f.read()\n",
    "\n",
    "question = \"List all genes with their molecular type\"\n",
    "graph_uri = \"https://www.brainkb.org/version01\"\n",
    "\n",
    "prompt = generate_sparql_prompt_from_yaml(yaml_data, question, graph_uri)\n",
    "print(prompt)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec554af7-e39b-488e-a5a2-602f94638d1e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
